### ISSUE TITLE
Issue with reconnect kernel using connectToKernel API

### ISSUE THREAD

**Issue by mahesh2013** [jupyter/jupyter-js-services#151]
When I try to reconnect a running kernel using connectToKernel API fails. When I use id with urlJoinEncode it seem to create kernel connection, but can't able to bind to web socket. If I don't use urlJoinEncode, then also it can't find a running kernel. Because of this I have to keep kernel connection info on memory and use that to send code to kernel gateway. I'm looking a way to use kernel id to reconnect running kernel.

Here is the sample code I used:
````
connectKernel: function (code) {
    console.log("Get list of Kernels");
    // Get a list of available kernels and connect to one.
    jupyter.listRunningKernels({
        baseUrl: baseUrl,
        ajaxSettings: ajaxSettings
    }).then(kernelModels => {
      console.log('kernelModels:', kernelModels );
      var id = utils.urlJoinEncode(kernelModels[0].id);
      console.log('Id:'+ id );
      jupyter.connectToKernel(id, {
          baseUrl: baseUrl,
          name: kernelModels[0].name,
          ajaxSettings: ajaxSettings
        }).then((kernel) => {
        console.log("#########Selected KERNEL FROM LIST##########");
        console.log(kernel.name);
        console.log('Executing sample code');
        console.log("CODE:" + code["code"]);

        var future = kernel.execute({ code: code["code"]  } );
        future.onDone = () => {
            console.log('Done execution.');
        };
        future.onIOPub = (msg) => {
            // print received messages
            console.log('Received message:', msg);
        };
      }).catch((error) => {
         console.error(error.message);
      });
    });
  }
````
Here is the output I get when I execute above (IP address masked)
````
Get list of Kernels
kernelModels: [ { name: 'python2',
    id: '7b4086b9-1b4e-40c8-a6d8-1c54773708a5-xceHGdlqsjQAKAANe3fy+cg+BpS90IeCUi+Ch9tzEJV8MJbLN6ZLaTEE8G/TiXOhCg/bgiSPgzGMym9JU4XbocYoWgRebg31dktKJvf0F+hpN1o47eFFXmxQVxbbeRz5ArRAwKdNMwzy1tcc4bN5yqNjmhrO2AtARGA4+5R33R9lSPjULgbl8v46+1ceysLyhNrRQKy89M7PaaZIfF+kR+gnl/RaWuKDvyMDqPztb2zPmL4XKLYGlWPUDyZZF5qjVacco0EEKXQG/H/sTlo0Ag==' } ]

Starting WebSocket: wss://XXXXX/jupyter/v1/api/kernels/7b4086b9-1b4e-40c8-a6d8-1c54773708a5-xceHGdlqsjQAKAANe3fy%252Bcg%252BBpS90IeCUi%252BCh9tzEJV8MJbLN6ZLaTEE8G/TiXOhCg/bgiSPgzGMym9JU4XbocYoWgRebg31dktKJvf0F%252BhpN1o47eFFXmxQVxbbeRz5ArRAwKdNMwzy1tcc4bN5yqNjmhrO2AtARGA4%252B5R33R9lSPjULgbl8v46%252B1ceysLyhNrRQKy89M7PaaZIfF%252BkR%252Bgnl/RaWuKDvyMDqPztb2zPmL4XKLYGlWPUDyZZF5qjVacco0EEKXQG/H/sTlo0Ag%253D%253D
#########Selected KERNEL FROM LIST##########
python2
Executing sample code
CODE:print('Executed on python Kernel')
Kernel:  (7b4086b9-1b4e-40c8-a6d8-1c54773708a5-xceHGdlqsjQAKAANe3fy%2Bcg%2BBpS90IeCUi%2BCh9tzEJV8MJbLN6ZLaTEE8G/TiXOhCg/bgiSPgzGMym9JU4XbocYoWgRebg31dktKJvf0F%2BhpN1o47eFFXmxQVxbbeRz5ArRAwKdNMwzy1tcc4bN5yqNjmhrO2AtARGA4%2B5R33R9lSPjULgbl8v46%2B1ceysLyhNrRQKy89M7PaaZIfF%2BkR%2Bgnl/RaWuKDvyMDqPztb2zPmL4XKLYGlWPUDyZZF5qjVacco0EEKXQG/H/sTlo0Ag%3D%3D)
Connection lost, reconnecting in 1 seconds.
Starting WebSocket: wss://XXXXX/jupyter/v1/api/kernels/7b4086b9-1b4e-40c8-a6d8-1c54773708a5-xceHGdlqsjQAKAANe3fy%252Bcg%252BBpS90IeCUi%252BCh9tzEJV8MJbLN6ZLaTEE8G/TiXOhCg/bgiSPgzGMym9JU4XbocYoWgRebg31dktKJvf0F%252BhpN1o47eFFXmxQVxbbeRz5ArRAwKdNMwzy1tcc4bN5yqNjmhrO2AtARGA4%252B5R33R9lSPjULgbl8v46%252B1ceysLyhNrRQKy89M7PaaZIfF%252BkR%252Bgnl/RaWuKDvyMDqPztb2zPmL4XKLYGlWPUDyZZF5qjVacco0EEKXQG/H/sTlo0Ag%253D%253D
Connection lost, reconnecting in 2 seconds.
Starting WebSocket: wss://XXXXX/jupyter/v1/api/kernels/7b4086b9-1b4e-40c8-a6d8-1c54773708a5-xceHGdlqsjQAKAANe3fy%252Bcg%252BBpS90IeCUi%252BCh9tzEJV8MJbLN6ZLaTEE8G/TiXOhCg/bgiSPgzGMym9JU4XbocYoWgRebg31dktKJvf0F%252BhpN1o47eFFXmxQVxbbeRz5ArRAwKdNMwzy1tcc4bN5yqNjmhrO2AtARGA4%252B5R33R9lSPjULgbl8v46%252B1ceysLyhNrRQKy89M7PaaZIfF%252BkR%252Bgnl/RaWuKDvyMDqPztb2zPmL4XKLYGlWPUDyZZF5qjVacco0EEKXQG/H/sTlo0Ag%253D%253D
Connection lost, reconnecting in 4 seconds.
````
I have also tried to run without urlJoinEncode, but in this case it fail to find a running Kernel.

**Comment by blink1073** [comment]
Hi @mahesh2013, the id should be just the `id` of the existing kernel, not url encoded:

snip:

```
connectKernel: function (code) {
    console.log("Get list of Kernels");
    // Get a list of available kernels and connect to one.
    jupyter.listRunningKernels({
        baseUrl: baseUrl,
        ajaxSettings: ajaxSettings
    }).then(kernelModels => {
      console.log('kernelModels:', kernelModels );
      var id = kernelModels[0].id;   // <-- Changed here.
      console.log('Id:'+ id );
      jupyter.connectToKernel(id, {
          baseUrl: baseUrl,
          name: kernelModels[0].name,
          ajaxSettings: ajaxSettings
        }).then((kernel) => {

**Comment by blink1073** [comment]
Assuming you are using the same notebook session, that should work between reconnects.

**Comment by mahesh2013** [comment]
Hi @blink1073 Yes, I have tried without url encoded same way as you have suggested above, but in that case I get "API request failed" error during connectToKernel like below:
````
Get list of Kernels
kernelModels: [ { name: 'python2',
    id: 'e2d31b4b-ab8f-4f4a-9668-bb9c5d650158-ifM+7C5UhFbYUqe3+Zut954Q44j4SmFCVPHFXIW/DLAeI6uTMVQFMwZ4BAlXf9zIe3UDpvMYrRD0rBvSkD27CXJtEm6R7Nk+nqinPA/K7fiTd4qAReS0J2HawNk2ANIpBLOhpYfjs/a4wCJXxipLR+fCET6dgvXtdnmqttVeOukhi3Pgasi+oE26//y67w8Nrga88RRB3oxzhlZboE7+TbMOpzZ0/6BxT9VPYuwFNO3850DxCDNyNne73w8Z2tIHrOSbGGXWdZH8MWYVWuPgBw==' } ]
Id:e2d31b4b-ab8f-4f4a-9668-bb9c5d650158-ifM%2B7C5UhFbYUqe3%2BZut954Q44j4SmFCVPHFXIW/DLAeI6uTMVQFMwZ4BAlXf9zIe3UDpvMYrRD0rBvSkD27CXJtEm6R7Nk%2BnqinPA/K7fiTd4qAReS0J2HawNk2ANIpBLOhpYfjs/a4wCJXxipLR%2BfCET6dgvXtdnmqttVeOukhi3Pgasi%2BoE26//y67w8Nrga88RRB3oxzhlZboE7%2BTbMOpzZ0/6BxT9VPYuwFNO3850DxCDNyNne73w8Z2tIHrOSbGGXWdZH8MWYVWuPgBw%3D%3D
API request failed (null):
No running kernel with id: e2d31b4b-ab8f-4f4a-9668-bb9c5d650158-ifM%2B7C5UhFbYUqe3%2BZut954Q44j4SmFCVPHFXIW/DLAeI6uTMVQFMwZ4BAlXf9zIe3UDpvMYrRD0rBvSkD27CXJtEm6R7Nk%2BnqinPA/K7fiTd4qAReS0J2HawNk2ANIpBLOhpYfjs/a4wCJXxipLR%2BfCET6dgvXtdnmqttVeOukhi3Pgasi%2BoE26//y67w8Nrga88RRB3oxzhlZboE7%2BTbMOpzZ0/6BxT9VPYuwFNO3850DxCDNyNne73w8Z2tIHrOSbGGXWdZH8MWYVWuPgBw%3D%3D
````

**Comment by blink1073** [comment]
Hmm, the uuid should only be 36 characters, I'm not sure why yours is coming back with extras, does it work if you use `id = id.slice(0, 36)` before passing it in to `connectToKernel`?

**Comment by parente** [comment]
@blink1073 I think @mahesh2013 is using a custom kernel provider that generates longer IDs. I think the problem here is that the kernel ID is URL encoded in many methods because of the use of `urlJoinEncode`, but not when `findById` is used.

**Comment by blink1073** [comment]
@mahesh2013, does making the following change to the code fix it?

```diff
diff --git a/src/kernel.ts b/src/kernel.ts
index aaac572..6a4fc06 100644
--- a/src/kernel.ts
+++ b/src/kernel.ts
@@ -145,7 +145,7 @@ function findKernelById(id: string, options?: IKernel.IOptions): Promise<IKernel
   let kernels = Private.runningKernels;
   for (let clientId in kernels) {
     let kernel = kernels[clientId];
-    if (kernel.id === id) {
+    if (utils.urlJoinEncode(kernel.id) === utils.urlJoinEncode(id)) {
       let result = { id: kernel.id, name: kernel.name };
       return Promise.resolve(result);
     }
```

**Comment by mahesh2013** [comment]
@blink1073, I have tried your suggested above change to findKernelById, but still I get same issue.
When I tried below change to  I can overcome this issue:
https://github.com/mahesh2013/jupyter-js-services/commit/9a44a492da3255a113d452e4fe469af8285b9320
Please check whether this change is acceptable, so that I can create a pull request for this.

**Comment by blink1073** [comment]
Thanks @mahesh2013, that looks good, and I will gladly accept the PR.

**Comment by blink1073** [comment]
Fixed by #157.
